kernel/src/hil/flash.rs

```diff
@@ -142,3 +164,111 @@ pub trait Client<F: Flash> {
     /// Flash erase complete.
     fn erase_complete(&self, error: Error);
 }
+
+/// A checked flash region
+pub struct Page {
+    pub address: u64,
+    pub length: u32,
+}
+
+/// A page of writeable persistent flash memory.
+///
+/// `W`: Should be the minimum number of bytes that can be written
+///      in an operation.
+/// `E`: Should be the minimum number of bytes that can be erased
+///      in an operation.
+pub trait Flash<const W: usize, const E: usize> {
+    /// Read data from flash into a buffer.
+    ///
+    /// This function will read data stored in flash at `address` and
+    /// `length` into the buffer `buf`.
```

Comment:
What if the buffer is smaller than the requested range? Will the buffer get filled multiple times?

What if the buffer is slightly larger than a page size? Does it make sense to require the driver to match the neighboring pages if the hardware only allows reading one page at a time?

Easy readout may be the point of this HIL, but I also think there's an intermediate level of complexity which separates write and erase size but doesn't care about comfortable reads like that.